core/app/models/spree/order.rb

```diff
@@ -418,23 +418,23 @@ def valid_credit_cards
     # Finalizes an in progress order after checkout is complete.
     # Called after transition to complete state when payments will have been processed
     def finalize!
-      # lock all adjustments (coupon promotions, etc.)
-      all_adjustments.each(&:finalize!)
-
-      # update payment and shipment(s) states, and save
-      updater.update_payment_state
-      shipments.each do |shipment|
-        shipment.update_state
-        shipment.finalize!
-      end
-
-      updater.update_shipment_state
-      save!
-      updater.run_hooks
+      Spree::Event.instrument 'order_finalize', order: self do
```

Comment:
I think we should have a constant for each event topic.
```ruby
Spree::EventTopics::ORDER_FINALIZE
#or 
Spree::EventTopics::Order::FINALIZE
```

Reply:
Yes, maybe. The benefit I can see for using constants is that we can kind of document all the events at the top of each class.  It would be easier to spot all of them. Do you have other benefits in mind? 

There's no hard constraint that would force devs to use constants, so the rule can be easily bypassed. A rule easily bypassable is usually not worth much.

AFAIK Rails codebase does not use any constant with ActiveSupport notifications, so maybe we can use the same direct approach without loosing much? Still pondering our options...

Reply:
I think you're right, we don't lose much by not having constants. However thinking from and extension stand point it would be easier to manage if constants were available.

Lets say I have many subscribers in my codebase defined with the static string `'order_finalize'` and one day down the road that topic name changes. When I run bundle update, all of sudden my code breaks.  If I were to define my subscribers with the constant, running bundle update wouldn't break a thing. In general I always think it's best to avoid magic values when possible

Reply:
👍 I agree constants are stronger than simple strings, but constant renaming and behavior changes may anyway break things in the future. I'm still thinking about other ways to make this more robust. Let's wait for some other feedback on this, thank you!

Reply:
> Lets say I have many subscribers in my codebase defined with the static string `'order_finalize'` and one day down the road that topic name changes. When I run bundle update, all of sudden my code breaks.

That's true, even if no one ensures us that someone just skipped the usage of the constant and used its content (the string) instead to subscribe to events. In that case, those stores would break in the same way. I imagine a way to change an event name could be:

```ruby
Spree::Event.instrument 'new_event_name', order: self do
  # Do stuff

  # Instrument the old event as well, so old subscribers will be happy
  Spree::Event.instrument 'old_event_name', order: self
end
```

We should also find a way in the Spree::Event#subscribe to deprecate events so subscribers to old events would be notified that it's time to change the name of the event.

My point is that if we decide to use constants but we still need to implement something similar cause we can't be sure devs used strings or constants while subscribing to event, I can't see a real advantage introducing them.

Reply:
I agree, I've come to a similar conclusion. It would make more sense for the extension that may subscribe many times to define the constant themselves at their discretion.